File: day-4/part-2.py
if __name__ == '__main__':
    with open('input.txt') as f:
        lines = f.readlines()
        lines = [line.strip() for line in lines]

        count = 0
        w = len(lines[0])
        h = len(lines)
        for y in range(1, h - 1):
            for x in range(1, w - 1):
                if lines[y][x] != 'A':
                    continue

                c1 = lines[y - 1][x - 1]
                c2 = lines[y + 1][x + 1]
                if (c1 == 'S' and c2 == 'M') or (c1 == 'M' and c2 == 'S'):
                    c1 = lines[y - 1][x + 1]
                    c2 = lines[y + 1][x - 1]
                    if (c1 == 'S' and c2 == 'M') or (c1 == 'M' and c2 == 'S'):
                        # print(x + 1, y + 1)
                        count += 1

        print(count)

File: day-2/src/bin/part-2.rs
fn main() -> Result<(), std::io::Error> {
    let input = std::fs::read_to_string("input.txt")?;
    let count = input
        .lines()
        .map(|line| {
            let levels = line
                .split_whitespace()
                .map(|x| x.parse::<i32>().unwrap())
                .collect::<Vec<_>>();

            if levels.len() < 2 {
                return true;
            }

            let is_safe = |levels: &[i32]| {
                if levels.len() < 2 {
                    return true;
                }

                let increasing = (levels[1] - levels[0]) >= 0;
                for pair in levels.windows(2) {
                    let diff: i32 = pair[1] - pair[0];
                    if (diff >= 0) != increasing {
                        return false;
                    }

                    if !(1..=3).contains(&diff.abs()) {
                        return false;
                    }
                }

                true
            };

            for i in 0..=levels.len() {
                if i == 0 {
                    if is_safe(&levels) {
                        return true;
                    }

                    continue;
                }

                let mut new_levels = levels.clone();
                new_levels.remove(i - 1);
                if is_safe(&new_levels) {
                    return true;
                }
            }

            false
        })
        .fold(0i32, |count, safe| count + if safe { 1 } else { 0 });

    println!("{}", count);
    Ok(())
}
